pkg/rbac/middleware.go

```diff
@@ -5,19 +5,64 @@ import (
 	"encoding/json"
 	"net/http"
 	"strings"
+	"time"
 
 	"github.com/stolostron/search-v2-api/pkg/config"
 	authv1 "k8s.io/api/authentication/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/klog/v2"
 )
 
+func (rc *RbacCache) GetUser(clientToken string, r *http.Request) *UserRbac {
+
+	// Iterate Users to find if any with a matching token.
+	for _, user := range rc.Users {
+		if user.DoesTokenExist(clientToken) {
+			return user
+		} else if user.uid == user.GetIDFromTokenReview(user.UserInfo) {
+			return user
+		} else {
+			return nil
+		}
+		//user does not exist:	return nil
+	}
+	return nil
+
+}
+
+func (ur *UserRbac) ValidateToken(clientToken string, r *http.Request) bool {
+	// Check the cached timestamp.
+	if !time.Now().After(ur.expiresAt) {
+		klog.V(5).Info("User token has not expired and is already validated with token review. Can continue with authorization..")
+		return true
+	} else {
+		// if it's been more than a minute re-validate
+		klog.V(5).Info("User token authentication over 1 minute, need to re-validate token")
+		//re-validate token and if the token is still valid we update the timestamp:
+		authenticated, _, _ := verifyToken(clientToken, r.Context())
+		if authenticated {
+			klog.V(5).Info("User token re-validated. Updating Timestamp.")
+			//update timestamp:
+			ur.UpdateTokenTime(clientToken)
+			//update expiration:
+			ur.SetExpTime(clientToken, ur.ValidatedTokens[clientToken])
+			// Return true if the token is valid.
+			return true
+		} else {
+			return false
+		}
+	}
+}
+
 //verifies token (userid) with the TokenReview:
-func Middleware() func(http.Handler) http.Handler {
+func Middleware(rc *RbacCache) func(http.Handler) http.Handler {
```

Comment:
Trying to simplify the implementation. Middleware() needs to do 2 things.
1. Get the token from the http.Request. (This is existing code.)
2. Validate the token. (It can be from cache or a new TokenRequest, it's not important for Middleware() )
Renaming to AuthenticationMiddleware() because we'll addd other middleware functions later.

```go
func AuthenticationMiddleware(cache *RbacCache) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			var clientToken string
			// Keep existing code.
			
			isValid, validationError := cache.ValidateToken(clientToken)
			if error != nil {
			    // Respond with error.
			} else if !isValid {
			    // Respond with token invalid.
			}
			
			klog.V(5).Info("User authentication successful!")
			next.ServeHTTP(w, r)
		}
	}
}
```

Then, we implement cache.ValidateToken(token string). 

``` go
func (ur *RbacCache) ValidateToken(clientToken string, r *http.Request) bool {
	// Try to find token in the cache.
	for _, user := range rc.Users {
		if user.DoesTokenExist(token) { 
			return user.isValidToken(token) // need to implement this.
		}
	}
	
	// This is a new token, call TokenReview and cache result.
}
```

```go
func (user *UserRbac) isValidToken(t string) bool {
  tokenExpiration := user.validatedTokens[t]
  // if not expired return true.
  // else call TokenReview, update cache expiration and return result.
}

Reply:
Just for clarification, in the ValidateToken function above, that should be `user *UserRbac` not `ur *RbacCache`, right?